crates/sparrow-execution/src/lib.rs

```diff
@@ -8,5 +8,228 @@
     clippy::undocumented_unsafe_blocks
 )]
 
-//! Implementations of the pipelines to be executed.
+//! Execution of physical plans.
+
+use arrow_array::RecordBatch;
+use arrow_schema::{DataType, Field, Schema, SchemaRef, TimeUnit};
+use hashbrown::HashMap;
+use sparrow_interfaces::{ReadConfig, Source};
+use sparrow_physical::StepId;
+use sparrow_transforms::TransformPipeline;
+use std::sync::Arc;
+
+use error_stack::ResultExt;
+use itertools::Itertools;
+use sparrow_scheduler::{InputHandles, WorkerPool};
+use uuid::Uuid;
+
+mod error;
+mod source_tasks;
+mod write_channel_pipeline;
+
+#[cfg(test)]
 mod tests;
+
+pub use error::*;
+
+use crate::source_tasks::SourceTasks;
+use crate::write_channel_pipeline::WriteChannelPipeline;
+
+pub struct PlanExecutor {
+    worker_pool: WorkerPool,
+    source_tasks: SourceTasks,
+    read_config: Arc<ReadConfig>,
+}
+
+fn result_type_to_output_schema(result_type: &DataType) -> SchemaRef {
+    let mut output_fields = vec![
+        Arc::new(Field::new(
+            "_time",
+            DataType::Timestamp(TimeUnit::Nanosecond, None),
+            false,
+        )),
+        // TODO: Convert key hash to key.
+        Arc::new(Field::new("_key_hash", DataType::UInt64, false)),
+    ];
+    match result_type {
+        DataType::Struct(fields) => {
+            output_fields.extend(fields.iter().cloned());
+        }
+        other => output_fields.push(Arc::new(Field::new("output", other.clone(), true))),
+    };
+    let output_schema = Schema::new(output_fields);
+    Arc::new(output_schema)
+}
+
+impl PlanExecutor {
+    /// Create the plan executor for the given plan.
+    ///
+    /// This creates the worker threads but does not inject inputs, meaning
+    /// the threads won't have anything to do until `run` is called.
+    pub fn try_new(
+        query_id: String,
+        plan: sparrow_physical::Plan,
+        sources: &HashMap<Uuid, Arc<dyn Source>>,
+        output: tokio::sync::mpsc::Sender<RecordBatch>,
+    ) -> error_stack::Result<Self, Error> {
+        let mut executor = PlanExecutor {
+            worker_pool: WorkerPool::new(query_id).change_context(Error::Creating)?,
+            source_tasks: SourceTasks::default(),
+            read_config: Arc::new(ReadConfig {
```

Comment:
TODO: to plumb these through?

Reply:
I can probably plumb it through pretty easily since it comes from the ExecutionConfig.

Reply:
Actually, I'll probably defer. I think we should discuss whether it makes more sense to just pass the `ExecutionOptions` everywhere, vs. having this separate `ReadConfig`.